Binder support for selectorless

crisbeto opened the pull request:
These changes update the template binder to account for selectorless. This is necessary for further compiler work. Includes the following commits:

### refactor(compiler): allow different kinds of directive matchers to be passed to binder
Updates the target binder to allow either a selector-based or selectorless matcher to be passed in. This will allow us to skip some of the overhead when matching directives to nodes.

### refactor(compiler): add flag to enable selectorless parsing 
Adds a flag to `parseTemplate` to enable selectorless.

### refactor(compiler): validate references in selectorless 
Adds the following validations to the selectorless template parsing:
* Local references with values are not allowed (e.g. `#foo="bar"`).
* Multiple local references with the same on a component or directive are not allowed.

### refactor(compiler): account for selectorless in template binder 
Updates the template binder to account for the new selectorless AST nodes. This is a prerequisite to supporting template type checking of the new syntax.

crisbeto committed:
refactor(compiler): allow different kinds of directive matchers to be passed to binder

Updates the target binder to allow either a selector-based or selectorless matcher to be passed in. This will allow us to skip some of the overhead when matching directives to nodes.

crisbeto committed:
refactor(compiler): add flag to enable selectorless parsing

Adds a flag to `parseTemplate` to enable selectorless.

crisbeto committed:
refactor(compiler): validate references in selectorless

Adds the follow validations to the selectorless template parsing:
* Local references with values are not allowed (e.g. `#foo="bar"`).
* Multiple local references with the same on a component or directive are not allowed.

crisbeto commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -159,13 +159,18 @@ export function findMatchingDirectivesAndPipes(template: string, directiveSelect
+/** Object used to match template nodes to directives. */
+export type DirectiveMatcher<DirectiveT extends DirectiveMeta> =
+  | SelectorMatcher<DirectiveT[]>
+  | SelectorlessMatcher<DirectiveT[]>;
```
For context: I started off with this being `SelectorlessMatcher<DirectiveT>`, but I realized that one name can actually correspond to multiple directives: the directive that was referenced and any host directives it may have.

atscott replied:
Do you anticipate `SelectorlessMatcher` ever being used in a way that doesn't return an array? If not, then would it make sense to update it to always have an array and remove the `null` return type? It seems a bit odd to me to have a special `null` when empty array seems to fit just as well as a "no matches" return value

```
 constructor(private registry: Record<string, T[]>) {}

  match(name: string): T[] {
    return this.registry[name] ?? [];
  }
```

Things become a bit easier like below in `visitDirective`:

```
  visitDirective(node: Directive): void {
    if (!(this.directiveMatcher instanceof SelectorlessMatcher)) {
      return;
    }
    const directives = this.directiveMatcher.match(node.name);
    this.trackSelectorlessBindings(node, directives);
  }
```
`visitComponent` has similar simplification when the `null` is removed.

crisbeto replied:
Hmm I can see it going either way. The array is convenient during the directive matching, but for the TCB generation logic we might have to check if any directives matching for which I'd prefer the falsy value. I'll keep it like this for the current PR and revisit it once the type checking is more fleshed out.

alxhub commented on packages/compiler/src/directive_matching.ts:
```diff
@@ -472,3 +472,11 @@ export class SelectorContext<T = any> {
 }
+
+export class SelectorlessMatcher<T = unknown> {
+  constructor(private registry: Record<string, T>) {}
```
Prefer `Map` to object record types.

crisbeto committed:
refactor(compiler): account for selectorless in template binder

Updates the template binder to account for the new selectorless AST nodes. This is a prerequisite to supporting template type checking of the new syntax.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 3d86b3ade7c62b117749310745e74ba8aa1b4cea.

The changes were merged into the following branches: main
